optimizely/marketing-website
# add webpack

This PR addresses adding webpack module loader into the marketing repo.  I know this implementation isn't a very "webpack" way, but I'm first trying to mimic our build behavior that was happening with the `concat` task to get everything working, and then go back and refactor. 

Issues:
- [x] ~~How to invoke sourcemaps in `grunt-webpack`~~ Addressed by adding `devtools: source-maps` in grunt config options for webpack task.  Forked `grunt-webpack` plugin and delete this option in everything except `dev` config.
- [x]~~How to create a `production` flag in `grunt-webpack` in order to specify files to be uglified by webpack.~~ Forked `grunt-webpack` allowing to pass templated grunt config.
- [x]~~Bundling vendor files into one file and making their globals available outside of the bundle.  Currently, I'm accomplishing this through the `script-loader`, which seems to concat all files into a string and call `eval` on it.~~ Libraries are required directly from Bower in `global.js` and bundles with all other global files.
- [x] ~~Mimicking `concat` behavior for the `namespacePages` target https://github.com/optimizely/marketing-website/blob/master/grunt/concat.js. i.e. separate bundles will be created for `page` and `layout` level JS and they will be output to `dist/assets/js` directory in the appropriate directory structure `layout/*.js` or `page/*.js`.  I think this can be accomplished by something similar to http://stackoverflow.com/questions/22078420/grunt-webpack-wildcard-on-entry/22715972#22715972, in combination with some sort of function that parses whether it is a page or layout and appends the correct directory structure.~~ (this could be improved by having one task for pages and layouts, currently it is split in two tasks due to output path)
- [x] ~~Currently `concat` is creating a banner and footer around JS bundles for catching errors in staging/production and creating a closure with the `window.optly.mrkt` global defined. May need to build a webpack loader that implements this functionality, or just continue to use concat.~~ Modified webpack BannerPlugin and required it in package.json from dtothefp github.
- [ ] Handlebars is being bundled more than once.  Once in global.js and additional times in page/layout js. Find a way to have handlebars only bundles once, or bundled globally and not included in the hbs-loader requires ( not merge blocking )
- [x] ~~Move JSHint into webpack.~~ All JSHint is happening through webpack except for the linting the test files. JSHint config for all tasks is still necessary as it is `require`d in the `webpack` grunt task.


### To Test

make sure to `rm -rf node_modules && npm cache clear && npm i`
`rm -rf bower_components && bower i`
run `grunt server`
- watch the build, that webpack compiles correctly. It will be halted on any JSHint errors.
- inspect your console to make sure there are no JS errors
- inspect the dist directory and see that all of the JS files have an associate source map file and that individual modules are wrapped in the closure transforming window and document to w and d, respectively.
- test a page like `/events` that uses a client side handlebars template. These are now being compiled with the webpack handlebars loader.
- in the `webpack` grunt config file fine the `sharedPreLoaders` array.  Append a `console.log(targetName)` to the end of the query string 
```js
loader: 'inject-filename-loader?inject=' + encodeURIComponent('var targetName = __filename.replace("website-guts/assets/js/", "");\n\nconsole.log(targetName)')
```
run the `grunt server` task again.  You will see all individual JS module file names logged.  This is what is used in error reporting to GA and debugging in production.
- look in `website-guts/assets/js/global.js` and see how the `require` functions bootstrap the JS in a particular order.  This order is important for internal dependencies
- play around with some es6 if you want
- throw an error in the JS and check that it maps back to the appropriate source file through the source map http://cl.ly/27173s2T3p00
- run grunt build and inspect that source map files no longer exist in the `dist` directory and that JS has been uglified
- check it out in staging and see that there are no errors in the console.  Also, inspect that the source has been uglified.

### Enhancements
- we can now write es6
- source maps for local JS debugging
- JSHint inside webpack
- elimination of handlebars and uglify task
- drastic reduction of code in concat and JSHint task
- Handlebars is no longer required in the vendor dependencies
- future elimination of globals substituted by `require`s
- JS files are watched by webpack which applies a cache. First build is slowish, but subsequent live changes with server running are fast.
- probably lots of other stuff I'm not thinking of

### Additional
- removed `yepnope` from `bower.json`. We deprecated this a long time ago
